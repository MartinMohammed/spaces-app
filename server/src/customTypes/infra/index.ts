/**
 * Types for CDK Application Infrastructure
 *
 * This file defines various types related to the CDK (AWS Cloud Development Kit) application and its infrastructure.
 * It includes enums for different stacks and CfnOutputs used within the application to reference AWS CloudFormation resources.
 */

/** Enum for Stacks Provided by the CDK Application */
export enum Stacks {
  DATA = "DataStack",
  LAMBDA = "LambdaStack",
  AUTH = "AuthStack",
  API = "ApiStack",
  UI_DEPLOYMENT = "UIDeploymentStack",
}

/** Enum for CfnOutputs Defined in the CDK Application */
export enum CfnOutputs {
  SPACE_USER_POOL_ID = "SpaceUserPoolId",
  SPACE_USER_POOL_CLIENT_ID = "SpaceUserPoolClientId",
  SPACE_IDENTITY_POOL_ID = "SpaceIdentityPoolId",
  SPACE_FINDER_URL = "SpaceFinderUrl",
  SPACE_FINDER_PHOTOS_BUCKET_NAME = "SpaceFinderPhotosBucketName",
}

/**
 * Stacks Enum:
 *
 * - DATA: Represents the Data Stack, which contains resources related to data storage and databases.
 * - LAMBDA: Represents the Lambda Stack, which contains AWS Lambda functions and related resources.
 * - AUTH: Represents the Auth Stack, responsible for configuring AWS Cognito User Pool and Identity Pool for user authentication.
 * - API: Represents the API Stack, which contains resources related to AWS API Gateway and API deployment.
 * - UI_DEPLOYMENT: Represents the API Stack, which contains resources related to Deploying the User interface (frontend)
 *
 * CfnOutputs Enum:
 *
 * - SPACE_USER_POOL_ID: Represents the CloudFormation Output key for the Cognito User Pool ID.
 * - SPACE_USER_POOL_CLIENT_ID: Represents the CloudFormation Output key for the Cognito User Pool Client ID.
 * - SPACE_IDENTITY_POOL_ID: Represents the CloudFormation Output key for the Cognito Identity Pool ID.
 *
 * Usage:
 * - These enums provide convenient and consistent references to the various stacks and CloudFormation outputs
 *   used within the CDK application's infrastructure code.
 * - They are used for identifying, fetching, or passing around resource names and IDs during the CDK application's deployment process.
 */
