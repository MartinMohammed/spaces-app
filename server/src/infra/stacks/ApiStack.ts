/** This stack will our ApiGateway (rest api) */
import { Stack, StackProps } from "aws-cdk-lib";
import {
  AuthorizationType,
  CognitoUserPoolsAuthorizer,
  Cors,
  LambdaIntegration,
  MethodOptions,
  ResourceOptions,
  RestApi,
} from "aws-cdk-lib/aws-apigateway";
import { IUserPool } from "aws-cdk-lib/aws-cognito";
import { Construct } from "constructs";
import { HttpMethod } from "aws-cdk-lib/aws-lambda";

interface ApiStackProps extends StackProps {
  spacesLambdaIntegration: LambdaIntegration;
  userPool: IUserPool;
}

export class ApiStack extends Stack {
  constructor(scope: Construct, id: string, props: ApiStackProps) {
    super(scope, id, props);

    const api = new RestApi(this, "SpacesApi");

    /**
     * Creates a Cognito User Pools Authorizer for API Gateway to handle authentication using Amazon Cognito User Pools.
     *
     * @param {Construct} scope - The construct representing the AWS CDK app.
     * @param {string} id - A logical ID that uniquely identifies the authorizer within the AWS CDK app.
     * @param {CognitoUserPoolsAuthorizerProps} props - Properties to configure the Cognito User Pools Authorizer.
     * @returns {CognitoUserPoolsAuthorizer} - An instance of the CognitoUserPoolsAuthorizer class representing the API Gateway authorizer.
     */
    const authorizer = new CognitoUserPoolsAuthorizer(
      this,
      "SpacesApiAuthorizer",
      {
        cognitoUserPools: [props.userPool],

        // Specifies the source of the JWT token in the API Gateway request header.
        identitySource: "method.request.header.Authorization",
      }
    );

    // Attach the Cognito User Pools Authorizer to the API Gateway.
    authorizer._attachToApi(api);

    const methodOptionsWithAuth: MethodOptions = {
      authorizationType: AuthorizationType.COGNITO,
      authorizer: {
        authorizerId: authorizer.authorizerId,
      },
    };

    const resourceOptionsWithCors: ResourceOptions = {
      defaultCorsPreflightOptions: {
        allowOrigins: Cors.ALL_ORIGINS,
        allowMethods: Cors.ALL_METHODS,
      },
    };

    // Add a api ressource (path) by hand.
    const spacesResource = api.root.addResource(
      "spaces",
      resourceOptionsWithCors
    );

    // One lambda for exact one resource.
    spacesResource.addMethod(
      HttpMethod.GET,
      props.spacesLambdaIntegration,
      methodOptionsWithAuth
    );
    spacesResource.addMethod(
      HttpMethod.POST,
      props.spacesLambdaIntegration,
      methodOptionsWithAuth
    );
    spacesResource.addMethod(
      HttpMethod.DELETE,
      props.spacesLambdaIntegration,
      methodOptionsWithAuth
    );
    spacesResource.addMethod(
      HttpMethod.PUT,
      props.spacesLambdaIntegration,
      methodOptionsWithAuth
    );
  }
}
