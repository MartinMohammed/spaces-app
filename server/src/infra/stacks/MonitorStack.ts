// Import necessary constructs and classes from AWS CDK library.
import { Duration, Stack, StackProps } from "aws-cdk-lib";
import { Alarm, Metric, Unit } from "aws-cdk-lib/aws-cloudwatch";
import { SnsAction } from "aws-cdk-lib/aws-cloudwatch-actions";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Topic } from "aws-cdk-lib/aws-sns";
import { LambdaSubscription } from "aws-cdk-lib/aws-sns-subscriptions";
import { Construct } from "constructs";
import path from "path";

// Define a custom stack named "MonitorStack" that extends the Stack class from AWS CDK.
export class MonitorStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    // Call the constructor of the base class (Stack) with the provided scope, id, and props.
    super(scope, id, props);

    // Invoked by SNS when CloudWatch Alarm is triggered and sends the alarm message to Slack Webhook
    const webhookLambda = new NodejsFunction(this, "WebhookLambda", {
      runtime: Runtime.NODEJS_18_X,
      environment: {
        SLACK_WEBHOOK_URL:
          "https://hooks.slack.com/services/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",
      },
      handler: "handler",
      entry: path.join(
        __dirname,
        "..",
        "..",
        "services",
        "monitor",
        "handler.ts"
      ),
    });

    // Create SNS topic - where events are published to and subscribed from.
    const alarmTopic = new Topic(this, "AlarmTopic", {
      displayName: "AlarmTopic",
      topicName: "AlarmTopic", // The name of the SNS topic instead of auto-generated physical id.
    });

    // Subscribe the Lambda function (webhookLambda) to the SNS topic (alarmTopic)
    alarmTopic.addSubscription(new LambdaSubscription(webhookLambda));

    // Create a CloudWatch alarm to monitor 4XX errors in the AWS API Gateway.

    // Specify the configuration for the alarm by creating a new instance of the Alarm class.
    const spacesApi4xxAlarm = new Alarm(this, "spacesAPI4xxAlarm", {
      // The metric for the alarm is a custom metric representing the count of 4XX errors.
      metric: new Metric({
        metricName: "4XXError", // The name of the metric to monitor.
        namespace: "AWS/ApiGateway", // The namespace where the metric resides (API Gateway in this case).
        period: Duration.minutes(1), // The time period for the metric data (1 minute).
        statistic: "Sum", // The statistical method to apply to the data (Sum).
        unit: Unit.COUNT, // The unit of the metric (Count).
        dimensionsMap: {
          ApiName: "SpacesApi", // The dimension representing the specific API (named "SpacesApi").
        },
      }),
      evaluationPeriods: 1, // The number of periods over which the alarm condition should be evaluated (1).
      threshold: 5, // The threshold value that triggers the alarm when exceeded (5).
      alarmName: "spacesAPI4xxAlarm", // The name (physical id) of the alarm.
    });

    // Create SNS action for the alarm to send messages to the topic.
    const topicAction = new SnsAction(alarmTopic);

    // Trigger the SNS action if the alarm fires.
    spacesApi4xxAlarm.addAlarmAction(topicAction);

    // Trigger the SNS action if the alarm returns from the breaching state into the OK state.
    spacesApi4xxAlarm.addOkAction(topicAction);
  }
}
