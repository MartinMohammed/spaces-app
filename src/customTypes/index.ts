/**
 * General Types for CDK Application
 *
 * This file defines general types related to the entire CDK (AWS Cloud Development Kit) application.
 * It includes an enum representing common HTTP methods used in the application's API endpoints.
 */

/** Enum for Common HTTP Methods */
export enum HTTPMethods {
  GET = "GET",
  POST = "POST",
  DELETE = "DELETE",
  PUT = "PUT",
}

/**
 * HTTPMethods Enum:
 *
 * - GET: Represents the HTTP GET method used for retrieving data from the server.
 * - POST: Represents the HTTP POST method used for creating new resources on the server.
 * - DELETE: Represents the HTTP DELETE method used for removing resources from the server.
 * - PUT: Represents the HTTP PUT method used for updating existing resources on the server.
 *
 * Usage:
 * - This enum provides a standardized representation of common HTTP methods used in the CDK application's API endpoints.
 * - It is helpful for maintaining consistency and readability in API endpoint definitions and interactions.
 * - Developers can use these values to indicate the intended action of API requests and responses throughout the application.
 * - For example, when defining API Gateway resources, this enum can be used to specify the corresponding HTTP method.
 */
