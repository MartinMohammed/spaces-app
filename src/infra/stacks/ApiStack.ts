/** This stack will our ApiGateway (rest api) */
import { Stack, StackProps } from "aws-cdk-lib";
import { LambdaIntegration, RestApi } from "aws-cdk-lib/aws-apigateway";
import { Construct } from "constructs";

interface ApiStackProps extends StackProps {
  spacesLambdaIntegration: LambdaIntegration;
}

export class ApiStack extends Stack {
  constructor(scope: Construct, id: string, props: ApiStackProps) {
    super(scope, id, props);

    const api = new RestApi(this, "SpacesApi");

    // Add a api ressource (path) by hand.
    const spacesResource = api.root.addResource("spaces");

    // One lambda for exact one resource.
    spacesResource.addMethod("GET", props.spacesLambdaIntegration);
    spacesResource.addMethod("POST", props.spacesLambdaIntegration);
    spacesResource.addMethod("DELETE", props.spacesLambdaIntegration);
    spacesResource.addMethod("PUT", props.spacesLambdaIntegration);
  }
}
