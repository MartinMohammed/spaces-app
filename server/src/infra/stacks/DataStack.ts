/**
 * Stack responsible for creating and managing the state resources, including S3 bucket and DynamoDB table.
 * This stack will contain the S3 bucket to store user-uploaded images and the DynamoDB table for space data storage.
 */
import { CfnOutput, Stack, StackProps } from "aws-cdk-lib";
import { AttributeType, ITable, Table } from "aws-cdk-lib/aws-dynamodb";
import { Construct } from "constructs";
import { getSuffixFromStack } from "../../Utils";
import {
  Bucket,
  HttpMethods,
  IBucket,
  ObjectOwnership,
} from "aws-cdk-lib/aws-s3";
import { CfnOutputs } from "../../customTypes/infra";

export class DataStack extends Stack {
  // Make the DynamoDB table accessible to outside.
  public readonly spacesTable: ITable;
  public readonly photosBucket: IBucket;

  /**
   * Constructs a new DataStack instance.
   *
   * @param {Construct} scope - The scope in which this stack is created.
   * @param {string} id - The ID of the stack.
   * @param {StackProps} props - Optional stack properties.
   */
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const suffix = getSuffixFromStack(this);

    /**
     * Bucket where the user uploaded images will be statically stored.
     */
    this.photosBucket = new Bucket(this, "SpaceFinderPhotos", {
      bucketName: `space-finder-photos-${suffix}`,
      // Effectively making the bucket publicly accessible.
      blockPublicAccess: {
        blockPublicAcls: false,
        blockPublicPolicy: false,
        ignorePublicAcls: false,
        restrictPublicBuckets: false,
      },
      // The uploading account will own the object.
      objectOwnership: ObjectOwnership.OBJECT_WRITER,
      publicReadAccess: true,
      cors: [
        {
          allowedMethods: [HttpMethods.HEAD, HttpMethods.GET, HttpMethods.PUT],
          allowedOrigins: ["*"],
        },
      ],
    });

    new CfnOutput(this, CfnOutputs.SPACE_FINDER_PHOTOS_BUCKET_NAME, {
      value: this.photosBucket.bucketName,
    });

    // DynamoDB Table
    this.spacesTable = new Table(this, "SpacesTable", {
      tableName: `SpaceTable-${suffix}`,
      partitionKey: {
        type: AttributeType.STRING,
        name: "id",
      },
    });
  }
}
