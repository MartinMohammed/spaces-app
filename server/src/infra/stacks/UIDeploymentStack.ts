import { CfnOutput, Stack, StackProps, aws_s3_deployment } from "aws-cdk-lib";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";
import { getSuffixFromStack } from "../../Utils";
import { join } from "path";
import { existsSync } from "fs";
import { BucketDeployment } from "aws-cdk-lib/aws-s3-deployment";
import { Distribution, OriginAccessIdentity } from "aws-cdk-lib/aws-cloudfront";
import { S3Origin } from "aws-cdk-lib/aws-cloudfront-origins";
import { CfnOutputs } from "../../customTypes/infra";

/**
 * Stack for Frontend Application Deployment
 *
 * This stack contains resources for deploying the frontend application of the CDK application.
 * It leverages AWS CloudFormation to create and manage the required resources.
 * The frontend application consists of static files that are hosted and served using AWS S3 and AWS CloudFront for improved performance and scalability.
 */
export class UIDeploymentStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    /**
     * contains the static files for our frontend application and is responsible for serving our app.
     */

    // Only deploy resources if the dist folder with the static build files exist.

    // Create an S3 bucket to store the static files of the frontend application.
    const deploymentBucket = new Bucket(this, "uiDeploymentBucket", {
      bucketName: `space-finder-frontend-${getSuffixFromStack(this)}`,
    });

    // Specify the directory where the frontend application's build files are located.
    const uiDir = join(__dirname, "..", "..", "..", "..", "client", "dist");

    // Check if the specified directory (uiDir) exists.
    if (!existsSync(uiDir)) {
      console.warn("UI directory not found: ", uiDir);
      return;
    }

    // Deploy the frontend application's build files to the S3 bucket.
    new BucketDeployment(this, "SpacesFinderDeployment", {
      destinationBucket: deploymentBucket,
      sources: [aws_s3_deployment.Source.asset(uiDir)],
    });

    // An origin access identity is a special CloudFront user that you can associate with Amazon S3 origins,
    //  so that you can secure all or just some of your Amazon S3 content. (enable it to read from the s3 bucket)
    const originIdentity = new OriginAccessIdentity(
      this,
      "OriginAccessIdentity"
    );

    // Grant read permissions for this bucket and its contents to the CloudFront origin access identity.
    deploymentBucket.grantRead(originIdentity);

    // Create a CloudFront distribution to serve the frontend application from the S3 bucket.
    const distribution = new Distribution(this, "SpacesFinderDistribution", {
      defaultRootObject: "index.html",
      defaultBehavior: {
        // Specify the origin for the CloudFront distribution, which is the S3 bucket.
        origin: new S3Origin(deploymentBucket, {
          // Associate the CloudFront distribution with the origin access identity to enhance security.
          originAccessIdentity: originIdentity,
        }),
      },
    });

    // Create a CloudFormation output to display the URL of the CloudFront distribution.
    new CfnOutput(this, CfnOutputs.SPACE_FINDER_URL, {
      value: distribution.distributionDomainName,
    });
  }
}
