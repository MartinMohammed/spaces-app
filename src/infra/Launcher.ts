/** Serves as the entry point of cdk - instantiates the cdk app and imports the stacks */
import { App } from "aws-cdk-lib";
import { DataStack } from "./stacks/DataStack";
import { LambdaStack } from "./stacks/LambdaStack";
import { ApiStack } from "./stacks/ApiStack";

const app = new App();
const dataStack = new DataStack(app, "DataStack");
const lambdaStack = new LambdaStack(app, "LambdaStack", {
  spacesTable: dataStack.spacesTable,
});
new ApiStack(app, "ApiStack", {
  spacesLambdaIntegration: lambdaStack.spacesLambdaIntegration,
});
