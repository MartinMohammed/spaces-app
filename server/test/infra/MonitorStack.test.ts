import { App } from "aws-cdk-lib";
import { MonitorStack } from "../../src/infra/stacks/MonitorStack";
import { Stacks } from "../../src/customTypes/infra";
import { Capture, Match, Template } from "aws-cdk-lib/assertions";

/**
 * Automated Tests for the Monitor Stack's Constructs
 *
 * This test suite checks the properties of the constructs defined within the Monitor Stack.
 * The Monitor Stack is a part of the infrastructure and contains resources like AWS Lambda function and SNS topic.
 * These tests ensure that the expected properties of the Lambda function and SNS topic are correctly configured
 * and created within the stack.
 */

// Create a test suite named "Check the construct within the Monitor Stack"
describe("Check the construct within the Monitor Stack", () => {
  // Define a variable to store the CloudFormation template of the Monitor Stack
  let monitorStackTemplate: Template;

  // Before running any tests, set up the necessary environment
  beforeAll(() => {
    // Create a new App instance for testing with the specified output directory
    const testApp = new App({
      outdir: "cdk.out",
    });

    // Instantiate the MonitorStack using the App and the stack identifier from the Stacks enum
    const monitorStack = new MonitorStack(testApp, Stacks.MONITOR);

    // Retrieve the CloudFormation template of the Monitor Stack
    // This method allows you to inspect the generated CloudFormation template as it would be deployed to AWS without actually deploying it.
    monitorStackTemplate = Template.fromStack(monitorStack);
  });

  // Test: Lambda properties
  test("Lambda properties: ", () => {
    /**
     * Check if the Lambda Construct was created within the stack with the expected properties.
     * The Lambda function is expected to have a specific 'Handler' and 'Runtime' configuration.
     */

    monitorStackTemplate.hasResourceProperties("AWS::Lambda::Function", {
      // The properties expected for the Lambda function construct
      Handler: "index.handler", // The entry point (handler) of the Lambda function
      Runtime: "nodejs18.x", // The runtime environment for the Lambda function
    });
  });

  // Test: SNS properties
  test("SNS properties: ", () => {
    /**
     * Check if the SNS Construct was created within the stack with the expected properties.
     * The SNS topic is expected to have a specific 'DisplayName' and 'TopicName'.
     */
    monitorStackTemplate.hasResourceProperties("AWS::SNS::Topic", {
      // The properties expected for the SNS topic construct
      DisplayName: "AlarmTopic", // The display name of the SNS topic
      TopicName: "AlarmTopic", // The name of the SNS topic
    });
  });
  // Test: SNS properties
  test("SNS subscription properties - with matchers: ", () => {
    // Assert that a resource of the given type and properties exists in the CloudFormation template
    monitorStackTemplate.hasResourceProperties(
      "AWS::SNS::Subscription",

      // The keys and their values (or matchers) must match exactly with the target.
      Match.objectEquals({
        Protocol: "lambda",
        TopicArn: {
          // contains
          Ref: Match.stringLikeRegexp("AlarmTopic"),
        },
        Endpoint: {
          "Fn::GetAtt": [Match.stringLikeRegexp("WebhookLambda"), "Arn"],
        },
      })
    );
  });
  // Test: SNS properties with exact values using AWS test matchers
  test("SNS subscription properties - with exact values: ", () => {
    // Get the set of matching resources of a given type and properties in the CloudFormation template.
    const snsTopicMatches =
      monitorStackTemplate.findResources("AWS::SNS::Topic");
    const snsTopicName = Object.keys(snsTopicMatches)[0];

    const lambdaMatches = monitorStackTemplate.findResources(
      "AWS::Lambda::Function"
    );
    const lambdaName = Object.keys(lambdaMatches)[0];

    // The keys and their values (or matchers) must match exactly with the target.
    Match.objectEquals({
      Protocol: "lambda",
      TopicArn: {
        // contains
        Ref: snsTopicName,
      },
      // The subscriber of the SNS topic.
      Endpoint: {
        "Fn::GetAtt": [lambdaName, "Arn"],
      },
    });
  });

  // Test case for "Alarm Actions" using CDK test captors
  test("Alarm Actions", () => {
    // Create a new instance of the Capture class to store the captured data
    const alarmActionsCapture = new Capture();

    // Populate alarmActionsCapture with the value of the AlarmActions entry.
    // Assuming monitorStackTemplate is an object representing the CloudFormation stack template,
    // the hasResourceProperties method is likely a custom utility to check if the template has the specified resource properties.
    monitorStackTemplate.hasResourceProperties("AWS::CloudWatch::Alarm", {
      AlarmActions: alarmActionsCapture,
    });

    // Use normal Jest API to test the object.
    // The expect function is used to perform assertions.
    // It calls the asArray method on alarmActionsCapture to get the captured values as an array.
    // The toEqual matcher is used to compare the captured array to the expected array provided in the test case.
    // The expected array contains an object with a single property "Ref" whose value is expected to match the regular expression /^AlarmTopic/.
    expect(alarmActionsCapture.asArray()).toEqual([
      {
        Ref: expect.stringMatching(/^AlarmTopic/),
      },
    ]);
  });

  /**
   * The problem with snapshot testing related to CDK is
   * that there are a lot of things that are auto-generated by the CDK.
   * So the test would not be resistant to changes, that is why we are trying to
   * test more specific resources instead of the entire template.
   */

  // CDK snapshot testing for the Monitor stack template
  test("Monitor stack snapshot", () => {
    // This test ensures that the CDK-generated template matches the most recent snapshot.
    // The `toJSON` method converts the CloudFormation stack template to a JSON representation.
    // Jest will create a snapshot of the template, and subsequent runs will compare the template against this snapshot.
    expect(monitorStackTemplate.toJSON()).toMatchSnapshot();
  });

  // CDK snapshot testing for Lambdas
  test("Lambdas snapshot testing", () => {
    // Find all resources in the Monitor stack template that are of type "AWS::Lambda::Function".
    const lambdaMatches = monitorStackTemplate.findResources(
      "AWS::Lambda::Function"
    );

    // This test ensures that the list of Lambda resources found in the template matches the most recent snapshot.
    // Jest will create a snapshot of the Lambda resources, and subsequent runs will compare them against this snapshot.
    expect(lambdaMatches).toMatchSnapshot();
  });

  // CDK snapshot testing for SNS Topics
  test("SNSTopic snapshot testing", () => {
    // Find all resources in the Monitor stack template that are of type "AWS::SNS::Topic".
    const snsTopicMatches =
      monitorStackTemplate.findResources("AWS::SNS::Topic");

    // This test ensures that the list of SNS Topics found in the template matches the most recent snapshot.
    // Jest will create a snapshot of the SNS Topics, and subsequent runs will compare them against this snapshot.
    expect(snsTopicMatches).toMatchSnapshot();
  });
});
