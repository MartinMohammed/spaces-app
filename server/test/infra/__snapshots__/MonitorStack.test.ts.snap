// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check the construct within the Monitor Stack Lambdas snapshot testing 1`] = `
{
  "WebhookLambda118EFE5D": {
    "DependsOn": [
      "WebhookLambdaServiceRoleDB3A5765",
    ],
    "Properties": {
      "Code": {
        "S3Bucket": {
          "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
        },
        "S3Key": "5aa294708d3ce42bd851a2a7fb9964e36effc03f62b82a99bf22d84f4372749e.zip",
      },
      "Environment": {
        "Variables": {
          "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          "SLACK_WEBHOOK_URL": "https://hooks.slack.com/services/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",
        },
      },
      "Handler": "index.handler",
      "Role": {
        "Fn::GetAtt": [
          "WebhookLambdaServiceRoleDB3A5765",
          "Arn",
        ],
      },
      "Runtime": "nodejs18.x",
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;

exports[`Check the construct within the Monitor Stack Monitor stack snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AlarmTopicD01E77F9": {
      "Properties": {
        "DisplayName": "AlarmTopic",
        "TopicName": "AlarmTopic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "WebhookLambda118EFE5D": {
      "DependsOn": [
        "WebhookLambdaServiceRoleDB3A5765",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5aa294708d3ce42bd851a2a7fb9964e36effc03f62b82a99bf22d84f4372749e.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "SLACK_WEBHOOK_URL": "https://hooks.slack.com/services/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",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "WebhookLambdaServiceRoleDB3A5765",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "WebhookLambdaAlarmTopic625AB1F9": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "WebhookLambda118EFE5D",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "AlarmTopicD01E77F9",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "WebhookLambdaAllowInvokeMonitorStackAlarmTopic9F999C82D8D0AAF4": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "WebhookLambda118EFE5D",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "AlarmTopicD01E77F9",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "WebhookLambdaServiceRoleDB3A5765": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "spacesAPI4xxAlarm9E652253": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "AlarmTopicD01E77F9",
          },
        ],
        "AlarmName": "spacesAPI4xxAlarm",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "ApiName",
            "Value": "SpacesApi",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "4XXError",
        "Namespace": "AWS/ApiGateway",
        "OKActions": [
          {
            "Ref": "AlarmTopicD01E77F9",
          },
        ],
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 5,
        "Unit": "Count",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Check the construct within the Monitor Stack SNSTopic snapshot testing 1`] = `
{
  "AlarmTopicD01E77F9": {
    "Properties": {
      "DisplayName": "AlarmTopic",
      "TopicName": "AlarmTopic",
    },
    "Type": "AWS::SNS::Topic",
  },
}
`;
